File: 0121_best_time_to_buy_and_sell_stock/max_profit.js
/**
 * Time:    O(n)
 * Space:   O(1)
 * @param   {number[]} prices
 * @return  {number}
 */
const maxProfit = prices => {
  let minPrice = Number.MAX_SAFE_INTEGER
  return prices.reduce((maxProfit, e) => {
    if (e < minPrice) minPrice = e
    else if (e - minPrice > maxProfit) maxProfit = e - minPrice
    return maxProfit
  }, 0)
}

File: 0121_best_time_to_buy_and_sell_stock/max_profit.cpp
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
int max_profit(vector<int> &prices)
{
    int min_price = INT_MAX;
    int max_profit = 0;
    for (auto i : prices)
    {
        if (i < min_price)
            min_price = i;
        else if (i - min_price > max_profit)
            max_profit = i - min_price;
    }
    return max_profit;
}